### DISCUSSION TITLE
Message passing IndexError: index out of range in self

### DISCUSSION THREAD

**Discussion by YunYunY** [pyg-team/pytorch_geometric#2966] [Q&A] [answered]
Hi,
I am use the code gcn.py in the example with an extra max pool layer. Because I am working on the classification task for the whole graph. Every graph has different number of nodes. 

But I got an error message after few steps training
```
 File "/lib/python3.6/site-packages/torch_geometric/nn/conv/message_passing.py", line 158, in __collect__
    j if arg[-2:] == '_j' else i)
  File "/lib/python3.6/site-packages/torch_geometric/nn/conv/message_passing.py", line 127, in __lift__
    return src.index_select(self.node_dim, index)
IndexError: index out of range in self
```

The model structure

```

class GCN(torch.nn.Module):
    def __init__(self, args):
        super(GCN, self).__init__()
        num_feature = args.nodal
        self.conv1 = GCNConv(num_feature, 16, cached=True,
                             normalize=not args.use_gdc)
        self.conv2 = GCNConv(16, args.num_classes, cached=True,
                             normalize=not args.use_gdc)
        self.lin1 = torch.nn.Linear(args.num_classes, 256)
        self.lin2 = torch.nn.Linear(256, args.num_classes)

        # self.conv1 = ChebConv(data.num_features, 16, K=2)
        # self.conv2 = ChebConv(16, data.num_features, K=2)

    def forward(self, data):
        batch = data.batch
        x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr
  
        
        x = F.relu(self.conv1(x, edge_index, edge_weight))
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index, edge_weight)
        x = global_max_pool(x, batch) 
     
        # return x
        return F.log_softmax(x, dim=1)
```

If I check the data size in the forward by 
```
print(x.shape)
print(edge_index)
print(edge_weight)
```
I got
```
torch.Size([6, 2048])
torch.Size([2, 12])
torch.Size([12])
Epoch: [0][0], lr: 0.00010      Time 0.024 (0.024)      Data 0.019 (0.019)    T                                                                                                                                                                 otalLoss 4.4661 (4.4661)        Prec@1 0.000 (0.000)    Prec@5 0.000 (0.000)
torch.Size([16, 2048])
torch.Size([2, 96])
torch.Size([96])
Epoch: [0][1], lr: 0.00010      Time 0.004 (0.014)      Data 0.001 (0.010)    T                                                                                                                                                                 otalLoss 4.4662 (4.4662)        Prec@1 0.000 (0.000)    Prec@5 0.000 (0.000)
torch.Size([8, 2048])
torch.Size([2, 20])
torch.Size([20])
Epoch: [0][2], lr: 0.00010      Time 0.005 (0.011)      Data 0.003 (0.008)    T                                                                                                                                                                 otalLoss 4.4659 (4.4661)        Prec@1 0.000 (0.000)    Prec@5 0.000 (0.000)
torch.Size([7, 2048])
torch.Size([2, 16])
torch.Size([16])
Epoch: [0][3], lr: 0.00010      Time 0.005 (0.009)      Data 0.003 (0.006)    T                                                                                                                                                                 otalLoss 4.4656 (4.4660)        Prec@1 0.000 (0.000)    Prec@5 0.000 (0.000)
torch.Size([5, 2048])
torch.Size([2, 7])
torch.Size([7])

```
Then it follows the error message. 

Because the batch_size is 1, for every step the model output and target shapes are 
```
torch.Size([1, 87])
torch.Size([1])
torch.Size([1, 87])
torch.Size([1])
torch.Size([1, 87])
torch.Size([1])
torch.Size([1, 87])
torch.Size([1])
```
The criterion is `torch.nn.NLLLoss()`

Since the model is training without errors for the first few steps, I guess it's not related to the data preparation? But I can't tell what's wrong with the last step before the error that might cause an issue.

Thank you very much for any help. 

------------------------------------------------------------------------------------------------------------------
I am tracing back from the source code to check the errors. But I kind of get lost. May I ask where is `src.index_select` source code located in the `torch_geometric/nn/conv/message_passing.py`? Thank you very much.

**Comment by wsad1** [comment, marked as answer]
1. Firstly, `cached` should only be set to `true` in a transductive setting, since you are feeding in different graphs that is not the case. Not sure if that might be causing this error though. Could you please paste the full traceback, so we can see which line in `forward` is causing this error.
2. In the latest version of message passing the `src.index_select` in located [here](https://github.com/rusty1s/pytorch_geometric/blob/5bcfece72506eab34a912f8e28fc433a87e1a747/torch_geometric/nn/conv/message_passing.py#L129). Also you can find the `index_select` documentation [here](https://pytorch.org/docs/stable/generated/torch.index_select.html).

**Reply by rusty1s** [reply]
Thanks @wsad1 . I agree that this error might be related to using `cached`.

**Reply by YunYunY** [reply]
Thank you very much. Setting cached to False solved the issue.

I have a following up question regarding the max_pool layer in the same code. I'm implementing `x = global_max_pool(x, batch) ` over all the nodes in a graph. But what if I want to apply max_pool over part of the nodes? 
Say I have node index [0, 1, 2, 3, 4, 5]. I want to apply max_pool over part 1 that has nodes [0, 1, 2] and max_pool over part 2 that has nodes [3, 4, 5]. What I am thinking is to somehow pass the partition index into forward and apply max_pool after `x = self.conv2(x, edge_index, edge_weight)`. The partition index is generated when preparing `torch_geometric.data.Dataset`. Is there a way to pass this index with `Data` in the `process` function? After passing the index, how could I apply the max_pool on part of the graph? 

Thank you very much.

**Reply by wsad1** [reply]
> After passing the index, how could I apply the max_pool on part of the graph?

In the example above with node index [0, 1, 2, 3, 4, 5] the partition index is 2, and max pool is applied over nodes [0, 1, 2] and [3, 4, 5] separately. One thing you could do is to create a variable `partition = [0, 0, 0, 1, 1, 1]` using the partition index and pass it to max pool `x = global_max_pool(x, batch =  partition)` 

> Is there a way to pass this index with Data in the process function?

You could create `partition` in the process function and pass it to `Data` .
`data = Data(edge_index=edge_index, edge_attr=edge_attr, y=y, partition = partition)`. In fact you can extend Data object with any attribute you like, as described [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Data).
